servlet/src/main/java/com/linecorp/armeria/server/servlet/ServletHttpRequest.java

```diff
@@ -0,0 +1,768 @@
+import java.util.Collections;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.servlet.AsyncContext;
+import javax.servlet.DispatcherType;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.SessionTrackingMode;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpUpgradeHandler;
+import javax.servlet.http.Part;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.servlet.util.LinkedMultiValueMap;
+import com.linecorp.armeria.server.servlet.util.ServletUtil;
+
+import io.netty.buffer.Unpooled;
+
+/**
+ * The servlet request.
+ */
+public class ServletHttpRequest implements HttpServletRequest {
+    private static final int HTTPS_PORT = 443;
+    private static final int HTTP_PORT = 80;
+    private static final String HTTPS = "https";
+    private static final String HTTP = "http";
+    private static final String POST = "POST";
+
+    private static final Logger logger = LoggerFactory.getLogger(ServletHttpRequest.class);
+    private static final Locale[] DEFAULT_LOCALS = { Locale.getDefault() };
+    private static final String RFC1123_DATE = "EEE, dd MMM yyyy HH:mm:ss zzz";
+    private static final SimpleDateFormat[] FORMATS_TEMPLATE = {
+            new SimpleDateFormat(RFC1123_DATE, Locale.ENGLISH),
+            new SimpleDateFormat("EEEEEE, dd-MMM-yy HH:mm:ss zzz", Locale.ENGLISH),
+            new SimpleDateFormat("EEE MMMM d HH:mm:ss yyyy", Locale.ENGLISH)
+    };
+
+    private final ServiceRequestContext serviceRequestContext;
+    private final DefaultServletContext servletContext;
+    private final AggregatedHttpRequest httpRequest;
+    private final DefaultServletInputStream inputStream = new DefaultServletInputStream();
+    private final Map<String, Object> attributeMap = new ConcurrentHashMap<>(16);
+
+    private LinkedMultiValueMap<String, String> parameterMap = new LinkedMultiValueMap<>();
+    private List<Part> fileUploadList = new ArrayList<>();
+    private Boolean asyncSupportedFlag = true;
+
+    @Nullable
+    private String servletPath;
+    @Nullable
+    private String queryString;
+    @Nullable
+    private String pathInfo;
+    @Nullable
+    private String requestURI;
+    @Nullable
+    private String characterEncoding;
```

Comment:
Let's make this variable as final because we can extract this value when the `ServletHttpRequest` is initialized.
```java
final Charset charset = httpRequest.headers().contentType().charset();
if (charset != null) {
    ...
} else {
    ...
}
```

Reply:
Please consider all other fields if we can do the same approach. 😉